internal/cmd/operator-sdk/generate/internal/relatedimages.go

```diff
@@ -0,0 +1,151 @@
+		seenRelatedImages:       sets.NewString(),
+	}
+
+	for _, deployment := range manifestCol.Deployments {
+		containers := append(deployment.Spec.Template.Spec.Containers, deployment.Spec.Template.Spec.InitContainers...)
+		for _, container := range containers {
+			// containerRef can just be the deployment if there's only one container
+			// otherwise we need {{ deployment.Name }}-{{ container.Name }}
+			containerRef := deployment.Name
+			if len(containers) > 1 {
+				containerRef += "-" + container.Name
+			}
+
+			if err := col.collectFromEnvironment(containerRef, container.Env); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return col.collectedRelatedImages(), nil
+}
+
+const relatedImagePrefix = "RELATED_IMAGE_"
+
+type relatedImage struct {
+	name         string
+	imageRef     string
+	containerRef string // If 1 container then {{deployment}} else {{deployment}}-{{container}}
+}
+
+type relatedImageCollector struct {
+	relatedImages           []*relatedImage
+	relatedImagesByName     map[string][]*relatedImage
+	relatedImagesByImageRef map[string][]*relatedImage
+	seenRelatedImages       sets.String
+}
+
+func (c *relatedImageCollector) collectFromEnvironment(containerRef string, env []corev1.EnvVar) error {
+	for _, envVar := range env {
+		if strings.HasPrefix(envVar.Name, relatedImagePrefix) {
+			if envVar.ValueFrom != nil {
+				return fmt.Errorf("related images with valueFrom field unsupported, found in %s`", envVar.Name)
+			}
+
+			name := c.formatName(envVar.Name)
+			c.collect(name, envVar.Value, containerRef)
+		}
+	}
+
+	return nil
+}
+
+func (c *relatedImageCollector) collect(name, imageRef, containerRef string) {
+	// Don't add exact duplicates (same name and image)
+	key := name + "-" + imageRef
+	if c.seenRelatedImages.Has(key) {
+		return
+	}
+	c.seenRelatedImages.Insert(key)
+
+	relatedImg := relatedImage{
+		name:         name,
+		imageRef:     imageRef,
+		containerRef: containerRef,
+	}
+
+	c.relatedImages = append(c.relatedImages, &relatedImg)
+	if relatedImages, ok := c.relatedImagesByName[name]; ok {
+		c.relatedImagesByName[name] = append(relatedImages, &relatedImg)
+	} else {
+		c.relatedImagesByName[name] = []*relatedImage{&relatedImg}
+	}
+
+	if relatedImages, ok := c.relatedImagesByImageRef[imageRef]; ok {
+		c.relatedImagesByImageRef[imageRef] = append(relatedImages, &relatedImg)
+	} else {
+		c.relatedImagesByImageRef[imageRef] = []*relatedImage{&relatedImg}
+	}
+}
+
+func (c *relatedImageCollector) collectedRelatedImages() []operatorsv1alpha1.RelatedImage {
+	final := make([]operatorsv1alpha1.RelatedImage, 0, len(c.relatedImages))
+
+	for _, relatedImage := range c.relatedImages {
+		name := relatedImage.name
+
+		// Prefix the name with the containerRef on name collisions.
+		if len(c.relatedImagesByName[relatedImage.name]) > 1 {
+			name = relatedImage.containerRef + "-" + name
+		}
+
+		// Only add the related image to the final list if its the first occurrence of an image.
```

Comment:
nit: only fix if there are other changes required.
```suggestion
		// Only add the related image to the final list if it's the first occurrence of an image.
```